### ISSUE TITLE
Local camera not working locally

### ISSUE THREAD

**Issue by weiss** [jitsi/jitsi#540]
I'm running Jitsi 2.11.5611 installed from your repository on Debian stretch.  `Options` &rarr; `Video` lists and enables the local camera, but the picture remains black and I see the following message:

```
SEVERE: [96] org.jitsi.impl.neomedia.codec.video.HFlip.log() avfilter_graph_config: -22,
             format AVFRAME, 640x480, class org.jitsi.impl.neomedia.codec.video.AVFrame,
             pixFmt 1, deviceSystemPixFmt 0x56595559
```

The same happens when enabling "Show local video" in a video call.  However, the remote side *does* see my video stream.

This is probably due to some change between 2.11.5602 and 2.11.5604: the former version works for me, the latter doesn't.

Not sure this is an actual bug or just an incompatibility with some Debian stretch component.  The issue described in https://community.jitsi.org/t/camera-not-working-with-jitsi-desktop/15038/ sounds similar, but I'm not sure it's the same thing.

**Comment by damencho** [comment]
2.11.5602 was been openh264 was introduced, so maybe something about hflip as the message says...

**Comment by traud** [comment]
Exactly the same error here on Ubuntu 19.10. For me, also incoming video was affected. LibJitisi tries to flip the picture horizontally, to create a mirror-like user experience in the video settings. LiJitsi flips the picture of the remote party, so right is right and left is left.

Error -22 [seems](http://ffmpeg.org/doxygen/trunk/error_8h_source.html) to be `INPUT_AND_OUTPUT_CHANGED`. Because I have no idea what that means (or why it happens), I applied the patch from within `resources/install/debian/patches/remove-hflip`.

**Comment by traud** [comment]
Mhm. This is not generic to all Debian/Ubuntu because Jitsi Desktop 2.11 works in Ubuntu 16.04 LTS (FFmpeg 2.8). There it still does a HFlip. Is there any documentation where that HFlip source code and/or parameters were coming from? Which is the main platform for Jitsi Desktop (Windows 8.1, Windows 10 Version xxyy, …) and which FFmpeg version is bundled for the other platforms exactly?

**Comment by damencho** [comment]
> Is there any documentation where that HFlip source code and/or parameters were coming from?

That is in libjitsi. https://github.com/jitsi/libjitsi/blob/master/src/org/jitsi/impl/neomedia/codec/video/HFlip.java#L26



> Which is the main platform for Jitsi Desktop (Windows 8.1, Windows 10 Version yymm, …)

There is no such thing as main platform, we try to support all of them. We had proceeded with openh264, but since then we had several bumps in the build process (as not having windows certificates to sign) and few bugs to be checked before proceeding with releasing new stable version. But currently there is no body actively working on the client and I'm doing my best to find time, and I do not have a good success on that :)
The ffmpeg used is normally in the docs: https://github.com/jitsi/jitsi-lgpl-dependencies/blob/master/README-native.md
This is the version used for those that we do static linking (windows and macosx), and we try to support latest two LTS/stable releases of Ubuntu/Debian ...

**Comment by traud** [comment]
Mhm. When I download the [latest nightly](https://download.jitsi.org/jitsi/nightly/msi/) of Jitsi Desktop (2.11.5623), and start that from the command prompt, I get the very same error message (and symptom) in Windows 10 Version 1809 as well (vanilla installation, x64).

However, looking at the time of change when OpenH264 was added, this should create much more threads and issue reports by users. Does it work for you?

> That is in libjitsi.

I know. What I meant is the original source, guide, tutorial, or documentation. That code looks terrible complicated. I wonder which FFmpeg version was the last one working with that.

**Comment by damencho** [comment]
> What I meant is the original source, guide, tutorial, or documentation.

There is no such.

Yes, the code is very complicated as it can do a lot of stuff...

The last ffmpeg we used for long time with x264 was FFmpeg 1.0.10. 
I think this is the last README with version before we started switching to openh264: https://github.com/jitsi/jitsi-gpl-dependencies/tree/master/src/native/ffmpeg

**Comment by traud** [comment]
I am not about x264 versus OpenH264. I am about HFlip and when/where it was last tested/worked. I face the same issue with VP8 and VP9. That big OpenH264 change introduced some change which might be the cause of this one. If I can approach that change, narrow it down, I might find the culprit.

**Comment by traud** [comment]
Asked differently: Дамян, you committed that change about HFlip. On which platform(s) did you test that change? The first step about an issue, is to reproduce and confirm it. I face that issue in Ubuntu 18.04 LTS (FFmpeg 3) which is very similar to Debian 9 (Stretch), the platform used by the original reporter. I do not build Jitsi Desktop myself but just use the latest nightly build. I face the same error message and symptoms in Windows 10, again with the latest nightly version. This is just the video preview in Menu → Tools → Options → Video → Devices. Does it work for you, on which platform?

**Comment by damencho** [comment]
I don't remember testing it. I think Ingo was testing that at that point ... I don't have linux machine at this moment to test.

**Comment by traud** [comment]
What is your current = main development environment? Then, I re-test all my issues in that, to ease reproducing them.

> I don’t have linux machine

This issue here affects _all_ platforms except those using FFmpeg 3.1 to FFmpeg 1.2. Therefore, Ubuntu 16.04 LTS with its FFmpeg 2.8 is not affected.

> Ingo was testing that at that point

Ahh, I thought, you created that code. However, I now see that Ingo created his own branch and you just merged that branch. Again, this affects all platforms (except Linux) because the bundled FFmpeg is version 3.4.

> code is very complicated as it can do a lot of stuff

A better answer would have been: ’I am not the author of that code. Perhaps, Ingo can join this discussion and provide you the puzzle pieces where he got that knowledge from.‘ FFmpeg has a tutorial about [filters](http://ffmpeg.org/ffmpeg-filters.html). To create a horizontal flip, normally just `hflip` is required on the command line. However, because here we use `hflip` programmatically, a source and a sink are required:
```
buffer → hflip → buffersink
```
That is called a filter graph which is explained in [chapter 4.1](http://ffmpeg.org/ffmpeg-filters.html#toc-Filtergraph-syntax-1). The parameters for `buffer` can be found in the mentioned FFmpeg documentation as well. The Jitsi team has chosen to go for the flat-string syntax. Furthermore, in this case here, the pixel format 1 is the source, which is mentioned in the error above as well: `pixFmt 1`. The command `ffmpeg -pix_fmts` shows that this is `yuyv422`. Because the very same pixel format should be in the sink, the filter graph was changed to:
```
buffer → hflip → format → buffersink
```
I created a small C app which does exactly the same as the Java Native Code in LibJitsi. That allowed me to get more messages: ‘The filters 'Parsed_buffer_0' and 'Parsed_hflip_1' do not have a common format and automatic conversion is disabled.‘ This is interesting, because the FFmpeg guide states ‘Libavfilter will automatically insert scale filters where format conversion is required.’ The wording `scale` is for historic reasons, as it does not only mean scaling but all possible format conversions nowadays. Because `hflip` does not allow `yuyv422` and does not create `yuyv422` as pixel format, `scales` must be inserted:
```
buffer → scale  → hflip → scale → format → buffersink
```
Because theese are not four filters anymore but six filters now, the last filter changes its ID from 4-1=3 to 6-1=5. Et voilá! However, the FFmpeg documentation contained one interesting word ‘automatically’. Why is this not done automatically:

> Jitsi Desktop 2.11.5602: OpenH264 was introduced

The actual cause was not commit [4c9c5cd](https://github.com/jitsi/libjitsi/commit/4c9c5cd7af37f42970d053fb2619f7d795863b99) but commit [d8338bc](https://github.com/jitsi/jitsi-lgpl-dependencies/commit/d8338bcc9aaaa7d3476e16ee8b18e3ea373bf408#diff-f7e6208934bba5ae184bbac6bd6be6d1). With that commit, not only the build system was changed but also the functionality was changed, as automatic conversion on filter graphs was disabled on default: `AVFILTER_AUTO_CONVERT_NONE`. Et voilá! However, why is Ubuntu 16.04 LTS not affected by this?

Because this was a software bug in FFmpeg, see commit https://github.com/FFmpeg/FFmpeg/commit/16ea0bca14730d71dd319b7dc148598275928cfe. Although the symbols existed since [FFmpeg 1.0](https://github.com/FFmpeg/FFmpeg/commit/440af105f2306d3c7b3b3f4d7530bab910d49cb9), git lost that change partly when a big branch [was merged](https://github.com/FFmpeg/FFmpeg/commit/ad60b3b1811f6fce23baf8cf081f469244004aae). Consequently, in FFmpeg 3.1 to FFmpeg 1.2, the symbol existed but `AVFILTER_AUTO_CONVERT_NONE` does nothing. Nothing! Consequently, in those versions, automatic conversion stays enabled. And I guess, Ingo was using Ubuntu 16.04 LTS as main development environment at that time, with its default FFmpeg 2.8. Therefore, he did not notice that he is going to kill the video preview in Jitsi Desktop.

**Comment by damencho** [comment]
> That code looks terrible complicated.

I was referring to your statement above. Libjitsi is the library used for the media. You can create one to one audio/video calls, but you can also create conference calls, where audio is mixed locally and where video can be relayed, form the focus to the rest of the participants (maybe this mode was disabled at some point in favor of using libjitsi). Also you can use in those calls jitsi-videobridge as a relay for the video calls.
That library is also used for running jitsi-videobridge v1, which is still the default on when installing jitsi-meet.
And of course device handling, different audio systems and different renderers for different platforms. 
This is what I meant by `code is very complicated as it can do a lot of stuff`. And yes, I'm not the author, but I had worked on it over the years and this is the product of many authors through the years (around 10 years I would say, not sure when it was created :))

Thank you for the complete description of that bug and digging into it!!!

**Comment by traud** [comment]
What is your current = main development environment? Then, I re-test all my future issues in that, to ease reproducing them.

> worked on it

I was not about LibJitsi in general but HFlip and/or the latest latest change to HFlip in particular. That code looks terrible complicated on the first glance. The problem was, I needed a starting point to understand that _line_ of code. Even a – I copied it from here, or Ingo did that, or it worked on platform x – would have helped. Thanks to a bit of luck, I then found that documentation at FFmpeg. Any constructive help was/is appreciated.

**Comment by damencho** [comment]
Well the history was that we used some internal API to do that and when moving to the new versions that API did not exist and I was not able to fix that, and then Ingo stepped in and helped with that part.

**Comment by traud** [comment]
Damian, what is your personal, current = main, development environment? Then, I re-test all my future issues in that, to ease reproducing them.

**Comment by damencho** [comment]
I'm on mac

**Comment by ibauersachs** [comment]
@traud  I lost track on what is going on where with all your pull requests, but might find some time next week. Would you mind creating a new issue here in the Jitsi repo that links to all other issues/PRs in all other repos for an easier follow-up?
Thanks!

**Comment by traud** [comment]
I am fixing several individual things, actually near to all Pull Request are at their own. I linked their Issue Reports via mentions/fixes/closes. Those Pull Requests which depend on other requests have a checkmark list (actually only that FFmpeg 4 report). Consequently, if you start with a Pull Request, you should have a good starter. There is no big picture. I just ran Jitsi Desktop on Ubuntu 19.10 and want to fix all obvious issues before the upcoming Ubuntu 20.04 LTS.

When it comes to this issue report here, only two Pull Requests are related:
jitsi/jitsi-lgpl-dependencies#11
jitsi/libjitsi#503
Although both fix the same, I would go for both as it does not hurt.

**Comment by traud** [comment]
> links to all other issues/PRs in all other repos for an easier follow-up

jitsi/jitsi-lgpl-dependencies#10 – updated that to provide an (chronological) overview.

**Comment by dinojr** [comment]
> Exactly the same error here on Ubuntu 19.10.

Where did you find of how did you manage to build a jitsi desktop for Ubuntu 19.10 ? The debs in the jitsi repos have some dependency problems (libavformat57 needed but Ubuntu ships with libavformat58 for instance).

**Comment by traud** [comment]
> How did you manage to build Jitsi Desktop for Ubuntu 19.10 [or Ubuntu 20.04 LTS]?

(a) If you do not need video, you can [override](http://serverfault.com/a/506993) those [dependencies](https://github.com/jitsi/jitsi/pull/620).
(b) If you need video, you have to compile the library for LGPL (the bridge to the dependency FFmpeg), LibJitsi, and Jitsi Desktop: [my steps](https://community.jitsi.org/t/18436/2). Those steps are rather old because many of my patches are merged by now; and they are not complete when it comes to the video codec H.264 because some of my patches are not merged, yet. If you need not VP9 but H.264 as video codec, ping me again, and I update that guide to the state of today.
